File: sample/src/main/java/io/techery/sample/DividerItemDecoration.java
package io.techery.sample;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.support.annotation.DrawableRes;
import android.support.v4.content.res.ResourcesCompat;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;

public class DividerItemDecoration extends RecyclerView.ItemDecoration {

	private Drawable mDivider;

	public DividerItemDecoration(Context context) {
		final TypedArray a = context.obtainStyledAttributes(new int [] { android.R.attr.listDivider });
		mDivider = a.getDrawable(0);
		a.recycle();
	}

	public DividerItemDecoration(Drawable divider) { mDivider = divider; }

	public DividerItemDecoration (Context context, @DrawableRes int drawableId) {
		mDivider = ResourcesCompat.getDrawable(context.getResources(), drawableId, null);
	}

	@Override
	public void getItemOffsets (Rect outRect, View view, RecyclerView parent, RecyclerView.State state) {
		super.getItemOffsets(outRect, view, parent, state);
		if (mDivider == null) return;
		if (parent.getChildAdapterPosition(view) < 1) return;

		if (getOrientation(parent) == LinearLayoutManager.VERTICAL) outRect.top = mDivider.getIntrinsicHeight();
		else outRect.left = mDivider.getIntrinsicWidth();
	}

	@Override
	public void onDrawOver(Canvas c, RecyclerView parent, RecyclerView.State state) {
		if (mDivider == null) { super.onDrawOver(c, parent, state); return; }

		if (getOrientation(parent) == LinearLayoutManager.VERTICAL) {
			final int left = parent.getPaddingLeft();
			final int right = parent.getWidth() - parent.getPaddingRight();
			final int childCount = parent.getChildCount();

			for (int i=1; i < childCount; i++) {
				final View child = parent.getChildAt(i);
				final RecyclerView.LayoutParams params = (RecyclerView.LayoutParams) child.getLayoutParams();
				final int size = mDivider.getIntrinsicHeight();
				final int top = child.getTop() - params.topMargin;
				final int bottom = top + size;
				mDivider.setBounds(left, top, right, bottom);
				mDivider.draw(c);
			}
		} else { //horizontal
			final int top = parent.getPaddingTop();
			final int bottom = parent.getHeight() - parent.getPaddingBottom();
			final int childCount = parent.getChildCount();

			for (int i=1; i < childCount; i++) {
				final View child = parent.getChildAt(i);
				final RecyclerView.LayoutParams params = (RecyclerView.LayoutParams) child.getLayoutParams();
				final int size = mDivider.getIntrinsicWidth();
				final int left = child.getLeft() - params.leftMargin;
				final int right = left + size;
				mDivider.setBounds(left, top, right, bottom);
				mDivider.draw(c);
			}
		}
	}

	private int getOrientation(RecyclerView parent) {
		if (parent.getLayoutManager() instanceof LinearLayoutManager) {
			LinearLayoutManager layoutManager = (LinearLayoutManager) parent.getLayoutManager();
			return layoutManager.getOrientation();
		} else throw new IllegalStateException("DividerItemDecoration can only be used with a LinearLayoutManager.");
	}
}

File: sample-x-kotlin/src/main/kotlin/io/erva/sample/DividerItemDecoration.kt
package io.erva.sample

import android.content.Context
import android.graphics.Canvas
import android.graphics.Rect
import android.graphics.drawable.Drawable
import android.view.View
import androidx.annotation.DrawableRes
import androidx.core.content.res.ResourcesCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

class DividerItemDecoration : RecyclerView.ItemDecoration {

    private var divider: Drawable? = null

    constructor(context: Context) {
        val a = context.obtainStyledAttributes(intArrayOf(android.R.attr.listDivider))
        divider = a.getDrawable(0)
        a.recycle()
    }

    constructor(divider: Drawable) {
        this.divider = divider
    }

    constructor(context: Context, @DrawableRes drawableId: Int) {
        divider = ResourcesCompat.getDrawable(context.resources, drawableId, null)
    }

    override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView, state: RecyclerView.State) {
        super.getItemOffsets(outRect, view, parent, state)
        if (divider == null) return
        if (parent.getChildAdapterPosition(view) < 1) return

        if (getOrientation(parent) == LinearLayoutManager.VERTICAL)
            outRect.top = divider!!.intrinsicHeight
        else
            outRect.left = divider!!.intrinsicWidth
    }

    override fun onDrawOver(c: Canvas, parent: RecyclerView, state: RecyclerView.State) {
        if (divider == null) {
            super.onDrawOver(c, parent, state)
            return
        }

        if (getOrientation(parent) == LinearLayoutManager.VERTICAL) {
            val left = parent.paddingLeft
            val right = parent.width - parent.paddingRight
            val childCount = parent.childCount

            for (i in 1..childCount - 1) {
                val child = parent.getChildAt(i)
                val params = child.layoutParams as RecyclerView.LayoutParams
                val size = divider!!.intrinsicHeight
                val top = child.top - params.topMargin
                val bottom = top + size
                divider!!.setBounds(left, top, right, bottom)
                divider!!.draw(c)
            }
        } else { //horizontal
            val top = parent.paddingTop
            val bottom = parent.height - parent.paddingBottom
            val childCount = parent.childCount

            for (i in 1..childCount - 1) {
                val child = parent.getChildAt(i)
                val params = child.layoutParams as RecyclerView.LayoutParams
                val size = divider!!.intrinsicWidth
                val left = child.left - params.leftMargin
                val right = left + size
                divider!!.setBounds(left, top, right, bottom)
                divider!!.draw(c)
            }
        }
    }

    private fun getOrientation(parent: RecyclerView): Int {
        if (parent.layoutManager is LinearLayoutManager) {
            val layoutManager = parent.layoutManager as LinearLayoutManager
            return layoutManager.orientation
        } else
            throw IllegalStateException("DividerItemDecoration can only be used with a LinearLayoutManager.")
    }
}